### ISSUE TITLE
gpiochi412 as example

### ISSUE THREAD

**Issue by helloyan** [kopera/erlang-gpio#3]
`root@OpenWrt:~# ls /sys/class/gpio/`
`export       gpio432      gpiochip412  unexport`

how can i wrote the code 

`(change_gw@OpenWrt)8> gpio:open_chip("/dev/gpiochip412").`
`{error,enoent}`

`gpio:open_chip("/dev/gpiochip0").`
`{ok,#Ref<0.172490437.1821769745.83712>}`

`gpio:open_lines(Chip, [1], [output], [0], "my-application").`
`{error,16}`

`gpio:open_lines(Chip, [1], [output], [0], "my-application").`
`{error,ebadf}`

**Comment by asabil** [comment]
can you check `ls /dev/gpiochip*` this library doesn't use the sysfs interface but rather the gpio chardev interface.

**Comment by helloyan** [comment]
`root@OpenWrt:~# ls /dev/gpiochip0`
`/dev/gpiochip0`

**Comment by asabil** [comment]
`ls /dev/gpiochip*` to list all gpiochips not `ls /dev/gpiochip0`

**Comment by helloyan** [comment]
there is only gpiochip0,but i got it worked

(change_gw@OpenWrt)1> {ok, Chip} = gpio:open_chip("/dev/gpiochip0").
{ok,#Ref<0.2828961874.1827012623.141335>}
(change_gw@OpenWrt)2> {ok,L3} = gpio:open_lines(Chip, [20], [output], [0], "my-application").
{ok,#Ref<0.2828961874.1827012623.141337>}
(change_gw@OpenWrt)3> gpio:write_lines(L3, {1}).
ok
(change_gw@OpenWrt)4> gpio:write_lines(L3, {0}).
ok
(change_gw@OpenWrt)5> gpio:write_lines(L3, {1}).
ok


root@OpenWrt:~# cat /sys/kernel/debug/gpio | grep 20
 gpio20  : out low  func0 2mA pull up
root@OpenWrt:~# cat /sys/kernel/debug/gpio | grep 20
 gpio20  : out high func0 2mA pull up
root@OpenWrt:~# cat /sys/kernel/debug/gpio | grep 20
 gpio20  : out low  func0 2mA pull up
root@OpenWrt:~# cat /sys/kernel/debug/gpio | grep 20
 gpio20  : out high func0 2mA pull up

**Comment by helloyan** [comment]
(change_gw@OpenWrt)6> gpio:open_line_events(Chip, [21], [input], [rising_edge], "my-application").
** exception error: bad argument
     in function  gpio:open_line_events_nif/5
        called as gpio:open_line_events_nif(#Ref<0.2159652147.1042677777.247046>,
                                            [21],
                                            [input],
                                            [rising_edge],
                                            "my-application")
     in call from gpio:open_line_events/5 (/usr/src/change_gw/_build/default/lib/gpio/src/gpio.erl, line 183)
(change_gw@OpenWrt)7> gpio:open_line_events(Chip, 21, [output], [rising_edge], "my-application").
{error,ebadf}

**Comment by asabil** [comment]
> (change_gw@OpenWrt)6> gpio:open_line_events(Chip, [21], [input], [rising_edge], "my-application").
** exception error: bad argument
in function gpio:open_line_events_nif/5
called as gpio:open_line_events_nif(#Ref<0.2159652147.1042677777.247046>,
[21],
[input],
[rising_edge],
"my-application")
in call from gpio:open_line_events/5 (/usr/src/change_gw/_build/default/lib/gpio/src/gpio.erl, line 183)

You can only open 1 pin for monitoring, the second argument should be `21` not `[21]`

> (change_gw@OpenWrt)7> gpio:open_line_events(Chip, 21, [output], [rising_edge], "my-application").
{error,ebadf}

Your shell was the owner of the `Chip` you opened using `gpio:open_chip/2`, however the shell crashed due to the bad argument error, which caused the chip to be closed. You can check https://beta.erlang.org/faq/problems.html#idm1149 (Section 9.15) to understand this.

**Comment by helloyan** [comment]
(change_gw@OpenWrt)1> {ok, Chip} = gpio:open_chip("/dev/gpiochip0").
{ok,#Ref<0.4206521758.3536453649.143557>}
(change_gw@OpenWrt)2>  gpio:open_line_events(Chip, 21, [output], [rising_edge], "my-application").
{ok,{#Ref<0.4206521758.3536453649.143559>,<0.4560.0>}}
(change_gw@OpenWrt)3>

It's alright at the moment.

![image](https://user-images.githubusercontent.com/1699033/228400413-314c7188-da62-444f-8918-1559c8f7206b.png)
It would be beneficial to update the documentation.


I have another question.
If I have multiple line_events open, how can I close them? According to the API, the function close_line_events(_) does not accept any arguments.

**Comment by asabil** [comment]
Right, sorry about the incorrect documentation, that was a bad copy/paste from open_lines/5.

As for closing line events, you should call `close_line_events/1` for each `open_line_events/5` passing the return value from it:

```erlang
{ok, LineEvents} = gpio:open_line_events(Chip, 21, [], [rising_edge, falling_edge], "my-application"),
ok = gpio:close_line_events(LineEvents).
```
